### DISCUSSION TITLE
Please update the experimental firmware for the hub to hub communication

### DISCUSSION THREAD

**Discussion by MichaelVerschaeve** [pybricks/support#983] [Q&A]
Hi, I'm using the hub to hub communication between a Spike and Inventor Hub with great success...
Thank you for this wonderful feature...

However I'd also like to use the icons library to display icons on my hub displays.
It seems that this feature was added later than the experimental firmware that is required for the hub to hub.
Pybricks tells me that the attribute is missing on the hub.display

Traceback (most recent call last):
  File "Blue.py", line 108, in <module>
AttributeError: 'LightMatrix' object has no attribute 'icon'

Any chance that the hub to hub expirimental firmware can be updated to the most recent version of pybricks or the experimental hub to hub communication can be moved to the beta firmware?

Kind regards,

Michael

**Comment by laurensvalk** [comment]
I made an initial attempt at rebasing the branch. When it finishing building in a little while will find the zip files here: https://github.com/pybricks/pybricks-micropython/actions/runs/4383373382/

Not everything may work, though.

**Reply by MichaelVerschaeve** [reply]
Works fine for me. Thanks a bunch, hartelijk bedankt !

**Reply by thehomelessguy** [reply]
Thanks so much for this. 

We are a first year FLL team and as such have gone through multiple robot redesigns. We have used pybricks to make technic hub versions of our robots. As part of our robot design presentation we rebuilt some old designs (slightly modified to use the technic hubs) and are actually going to have them do a brief 30 second synchronized dance to show off how our designs have advanced over the season. Basically going to line the bots up. Have them start driving when told and do the dance. So far have we have made technic hub versions of the advanced driving base the the droidbot E from primelessons (took some work as both designs rely heavily on the PrimeHub for structure). Kids have had fun doing the coding as well as building some fun signs for the robots to hold up at the end of the dance. 

I believe the kids are also going to have our PrimeHub bot play "Never going to give you up" during the dance (they are still of the age where they think rick rolling is fun).

**Comment by dlech** [comment]
FYI, we've made an alternate revision of the hub-to-hub communication if you want to try it out too. https://github.com/pybricks/pybricks-micropython/pull/158

**Reply by michaelwareman** [reply]
I have been trying to use the Hub to Hub (2 Robot Inventor hubs) experimental communications. I first started with the firmware mentioned this website -- https://pybricks.com/projects/tutorials/wireless/hub-to-hub/broadcast/ (2561), but one of the Hubs just shuts down. I then found the ticket with a different firmware (newer 2178). Same thing, Hub shuts down. I then found this ticket and at least the Hub does not immediately shut down, after a bit of time it times out (version 2761).  Also, when I press the right button a couple of times, on the send Hub - the code with this line:
 hub = InventorHub(broadcast_channel=1) the program stops with this error message:
Traceback (most recent call last):
  File "send_test.py", line 11, in <module>
OSError: [Errno 116] ETIMEDOUT
Line 11 is this line -- hub.light.on(Color.YELLOW)

or this error message:
Traceback (most recent call last):
  File "send_test.py", line 12, in <module>
OSError: [Errno 116] ETIMEDOUT
Line 12 is this line -- remote = Remote()

I was reading that for the Broadcast (sender??) on the Move Hub causes this error when still connected to the PyBrickes Code. So I shut both Hubs down and restarted them. Some thing happens, no action from the second Hub (Observe -- receiver??) when I press the right button on the Broadcast a couple of times the Hub stops running the program.

**Reply by dlech** [reply]
Are you using a remote control also or just the two hubs? I suspect it is the `Remote()` line that is timing out because no remote is available for connecting.

**Reply by michaelwareman** [reply]
Yes, I am using two Lego Inventor Hubs. I will comment the remote line. That would also mean that I need lines to see the button presses on the sending Hub, rather than, button presses from the remote. Which is what the sending code is looking for. Thank you I will adjust the code for both the Broadcast and Observe programs and see if that helps. 

As some background information I am working on a Johnny 5 robot. One hub would be the master and control two drive motors, two motors that raise and lower the upper and lower body segments, and the ultrasonic sensor for obstacle detection. The second Hub would control the arms, head turning, and other undetermined functions. How would I setup the Broadcast and Observe on lines so that both Hubs can send and receive information? It would be helpful if the second Hub could tell the master Hub information about arm and head motor positions.

**Reply by dlech** [reply]
Here's how I would probably do it...

The main hub should have:

```python
from pybricks.hubs import InventorHub
...

hub1 = InventorHub(broadcast_channel=1, observe_channels=[2])
...

# main loop
while True:
    hub1.ble.broadcast(requested_head_position, requested_left_arm_position, requested_right_arm_position)
    hub2_state = hub1.ble.observe(2)

    if hub2_state is None:
        # indicate loss of signal
        ...
    else:
        actual_head_position = hub2_state[0]
        actual_left_arm_position = hub2_state[1]
        actual_right_arm_position = hub2_state[2]

    ...
    wait(10)
```

and the other:

```python
from pybricks.hubs import InventorHub
...

hub2 = InventorHub(broadcast_channel=2, observe_channels=[1])
...

# main loop
while True:
    hub2.ble.broadcast(actual_head_position, actual_left_arm_position, actual_right_arm_position)
    hub1_state = hub2.ble.observe(1)

    if hub1_state is None:
        # indicate loss of signal
        ...
    else:
        requested_head_position = hub1_state[0]
        requested_left_arm_position = hub1_state[1]
        requested_right_arm_position = hub1_state[2]

    ...
    wait(10)
```

**Reply by michaelwareman** [reply]
Hello David:
Thank you for the code sample but it appears that I need a more explicit and detailed example to work from. Is there a more complete sample code that I could work from? If not, can you make sample program so that hub1 just observes in coming data, say just the left_arm_position and prints that value? Then on hub2 just broadcasts the motor position left_arm_position (on Port.F) that I would manually change?

It appears that I am having troubles assigning variables that would collect the data and then transmit that to the other hub. I also do not understand how many channels are being assigned and why for the broadcast_channels = 1 there is no [] brackets, but for the observe_channel = [2] there are [] brackets. 

Also, when I copy your code sample I get a syntax error for the else: statement. What is missing, Why is this happening?

**Reply by dlech** [reply]
> . I also do not understand how many channels are being assigned and why for the broadcast_channels = 1 there is no [] brackets, but for the observe_channel = [2] there are [] brackets.

Each hub can only broadcast on one channel but can observe multiple channels from multiple other hubs if you have more than 2 hubs. This is in why in `broadcast_channel=1`, `channel` is singular - there is only one channel - and in `observe_channels=[2]`, `channels` is plural - there can be more than one, and thus a list of 0 or more channels. For technical reasons, the latter is necessary to allocate memory to hold the observed data.

**Reply by dlech** [reply]
I haven't tested it yet, but I made the requested example here: https://github.com/pybricks/pybricks-micropython/pull/158#issuecomment-1537548360

**Reply by dlech** [reply]
> Also, when I copy your code sample I get a syntax error for the else: statement. What is missing, Why is this happening?

I just did a syntax check on it and there are no errors, so either you changed something or copy/paste didn't work right.

**Comment by michaelwareman** [comment]
Yes, I have been using the code from comment #158. This is what I have working so far:
Sender code: 
```python
from pybricks.hubs import InventorHub
from pybricks.pupdevices import Remote, Motor
from pybricks.tools import wait
from pybricks.parameters import Color, Button, Port

hub = hub = InventorHub(broadcast_channel=1, observe_channels=[1, 2])
motor = Motor(Port.F)

print("radio fw", hub.ble.version())

hub.light.on(Color.YELLOW)

while True:
    remote_target = motor.angle()
    hub.ble.broadcast(remote_target)
    print(remote_target)

    target_angle = hub.ble.observe(2)
    data = target_angle
    print(data)

    #if target_angle is not None:
    #    motor.run_target(75, data)
```
This code works -- if I move the left motor arm, the receiver hub (master hub in my previous comment) shows the the change in the motor.angle() value. My next step was to then take that value and decide how how I would like to move the  to some new value. I then added an if statement to the receiver code (master hub) works. Here is the output from the print (data) (-3,) -- left_arm_position  or (100) -- value sent from the Receiver code below. When I uncomment the If statement above I get this error message:
```
Traceback (most recent call last):
  File "sender.py", line 23, in <module>
TypeError: can't convert tuple to int
I tried changing the line  -- data = target_angle  -- to this -- data = target_angle[0] I get this error:
Traceback (most recent call last):
  File "sender.py", line 19, in <module>
TypeError: 'NoneType' object isn't subscriptable
```

What am I doing wrong?

Receiver code:
```python
from pybricks.hubs import InventorHub
from pybricks.pupdevices import Motor
from pybricks.tools import wait
from pybricks.parameters import Port, Color

hub = InventorHub(broadcast_channel=2, observe_channels=[1])
motor = Motor(Port.F)

print("radio fw", hub.ble.version())

hub.light.on(Color.GREEN)

while True:
    data = hub.ble.observe(1)
    print(data)

    if data is not None:
        target_angle = 100
        hub.ble.broadcast(target_angle)
```
This code works -- if I move the left motor arm, the receiver hub (master hub in my previous comment) shows the the change in the motor.angle() value. My next step was to then take that value and decide how how I would like to move the  to some new value. I then added an if statement to the receiver code (master hub) works. Here is the output from the print (data) -- (-3,) -- left_arm_position

**Reply by michaelwareman** [reply]
I figured it out! I moved these lines:
```python
data = target_angle
print(data)
```
To after the if statement like this:

```python
if target_angle is not None:
    data = target_angle[0]
    print(data)
    motor.run_target(75, data)
```

I guess I just needed to play a little longer. Thank you for your help.
